localstack/http/request.py

```diff
@@ -1,18 +1,106 @@
+from typing import TYPE_CHECKING, Dict, Mapping, Optional, Tuple, Union
+
+if TYPE_CHECKING:
+    from _typeshed.wsgi import WSGIEnvironment
 
 from werkzeug.datastructures import Headers
-from werkzeug.sansio.request import Request as _SansIORequest
-from werkzeug.utils import cached_property
+from werkzeug.wrappers.request import Request as WerkzeugRequest
+
+from localstack.utils import strings
+
+
+def dummy_wsgi_environment(
+    method: str = "GET",
+    path: str = "",
+    headers: Optional[Union[Dict, Headers]] = None,
+    body: Optional[Union[bytes, str]] = None,
+    scheme: str = "http",
+    root_path: str = "/",
+    query_string: Optional[str] = None,
+    remote_addr: Optional[str] = None,
+    server: Optional[Tuple[str, Optional[int]]] = None,
+    raw_uri: Optional[str] = None,
+) -> "WSGIEnvironment":
+    """
+    Creates a dummy WSGIEnvironment that represents a standalone sans-IO HTTP requests.
+
+    See https://wsgi.readthedocs.io/en/latest/definitions.html#standard-environ-keys
+
+    :param method: The HTTP request method (such as GET or POST)
+    :param path: The remainder of the request URL's path. This may be an empty string, if the
+        request URL targets the application root and does not have a trailing slash.
+    :param headers: optional HTTP headers
+    :param body: the body of the request
+    :param scheme: the scheme (http or https)
+    :param root_path: The initial portion of the request URL's path that corresponds to the
+        application object.
+    :param query_string: The portion of the request URL that follows the “?”, if any. May be
+        empty or absent.
+    :param remote_addr: The address making the request
+    :param server: The server (tuple of server name and port)
+    :param raw_uri: The original path that may contain url encoded path elements.
+    :return: A WSGIEnvironment dictionary
+    """
+
+    # Standard environ keys
+    environ = {
+        "REQUEST_METHOD": method,
+        "SCRIPT_NAME": root_path.rstrip("/"),
+        "PATH_INFO": path,
+    }
+
+    if query_string is not None:
+        environ["QUERY_STRING"] = query_string
+
+    if headers:
+        content_type = headers.get("Content-Type")
+        if content_type:
+            environ["CONTENT_TYPE"] = content_type
+
+    data = strings.to_bytes(body) if body else b""
+    environ["CONTENT_LENGTH"] = str(len(data))
+
+    environ["SERVER_PROTOCOL"] = "HTTP/1.1"
+
+    if server:
+        environ["SERVER_NAME"] = server[0]
+        if server[1]:
+            environ["SERVER_PORT"] = str(server[1])
+        else:
+            environ["SERVER_PORT"] = "80"
+    else:
+        environ["SERVER_NAME"] = "127.0.0.1"
+        environ["SERVER_PORT"] = "80"
+
+    if remote_addr:
+        environ["REMOTE_ADDR"] = remote_addr
+
+    if headers:
+        # TODO: set HTTP_ Variables (header fields), not needed currently because we set them
+        #  directly after ``Request`` construction.
```

Comment:
q / nit: Where is that? And maybe we should then move it in here within this PR?
If there's reasoning to keep the environment setting of the header fields where is currently is, please explain it a bit more in the comment here.

Reply:
it would make the function re-usable for other components that want to create a valid WSGI environment (which it isn't without the headers). i think i'm going to just add the functionality :-)